### DISCUSSION TITLE
Relabelling facets of a hexahedral mesh

### DISCUSSION THREAD

**Discussion by cscmaw** [firedrakeproject/firedrake#3048] [Firedrake support] [answered]
Hi,

Following some previous discussions on here I can relabel facets of a tetrahedral mesh using the 'HDiv Trace' space.
In the online documentation it says this space works with hexahedral elements but the code fails and warns that it
doesn't. 

Is there a space I can use to relabel facets of a hexahedral mesh?  

A minimal example that fails:

from firedrake import *
gmesh = UnitCubeMesh(3,3,3,hexahedral=True)
x,y,z = SpatialCoordinate( gmesh )
V0 = FunctionSpace( gmesh, "HDiv Trace", 0 )
f1 = Function(V0).interpolate( conditional(x < .01,1.0,0.0) )
f2 = Function(V0).interpolate( conditional(x > .99,1.0,0.0) )
mesh = RelabeledMesh( gmesh, [f1,f2], [1,2] )

I want to use my own hexahedral geometry data, but the error is the same.

Thanks,
Mark.

**Comment by ksagiyam** [comment, marked as answer]
This PR will allow for marking facets with Q2 space for hex: https://github.com/firedrakeproject/firedrake/pull/3049.

```
from firedrake import *
  

gmesh = UnitCubeMesh(3, 3, 3, hexahedral=True)
x,y,z = SpatialCoordinate(gmesh)
V0 = FunctionSpace(gmesh, "Q", 2)
f1 = Function(V0).interpolate(conditional(x < .01, 1.0, 0.0))
f2 = Function(V0).interpolate(conditional(x > .99, 1.0, 0.0))
mesh = RelabeledMesh(gmesh, [f1, f2], [1, 2])
```

**Reply by cscmaw** [reply]
Thanks, that is great.

Will it pull through to the Docker image , as that is what I use due to our local restrictions on installing software?

Best regards,
Mark.

**Reply by ksagiyam** [reply]
Yes, it should just work:
```
cd $VIRTUAL_ENV/firedrake/src/firedrake
git fetch origin
git checkout ksagiyam/relabeledmesh_hex_workaround
```
